Automate the bootimage bump process

aaradhak opened the pull request:
This bootimagebump job would automate a part of the bootimage bump process by running the plume-cosa script and creating a PR for the rhcos build that is currently done manually. 

JIRA: https://issues.redhat.com/browse/COS-2205

aaradhak committed:
Automate the bootimage bump process

This bootimagebump job would automate a part of the bootimage bump
process by running the plume-cosa script and creating a PR for the
rhcos build that is currently done manually.

JIRA: https://issues.redhat.com/browse/COS-2205

aaradhak commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,145 @@
+            }
+
+            stage('Create Pull Request') {
+                //if (shwrapCapture("git diff --exit-code") != 0){
```
Using git diff --exit-code seem to throw an issue here. 

13:55:10  + git diff --exit-code
13:55:10  warning: Not a git repository. Use --no-index to compare two paths outside a working tree
13:55:10  usage: git diff --no-index [<options>] <path> <path>

mike-nguyen replied:
This isn't in the `installer` directory so it is failing there.  You can try `git -C installer diff --exit-code`

aaradhak replied:
With the git -C installer diff --exit-code , it seems to print out the output of git diff and then just exit the code with a failure, not quite evident why it is exiting after that check.

aaradhak commented:
Posting the PR opened with this job for reference - https://github.com/openshift/installer/pull/9455
(The indentation of the commit message needs to be addressed)

Jenkins job - https://jenkins-fedora-coreos-pipeline.apps.ocp.stg.fedoraproject.org/job/bootimagebump/107/console

aaradhak commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,146 @@
+        durabilityHint('PERFORMANCE_OPTIMIZED')
+    ])
+
+    COSA_IMAGE = "quay.io/coreos-assembler/coreos-assembler:rhcos-${params.STREAM}"
```
Actually, I think we can remove COSA_IMAGE here, it is not needed anymore. Missed to remove this duplicate.

aaradhak commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,146 @@
+    COSA_IMAGE = "quay.io/coreos-assembler/coreos-assembler:rhcos-${params.STREAM}"
+    RHCOS_METADATA_FILE = "data/data/coreos/rhcos.json"
+    PR_BRANCH = "bootimage-bump-${params.BUILD_VERSION}"
+    RELEASE_BRANCH = "release-${params.STREAM.split('-')[0]}"
```
I see, in that case, this might need to be change to cater the RHEL stream format

aaradhak replied:
I have added the condition to address the RHEL format as well

aaradhak commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,146 @@
+                            --distro rhcos \
+                            --no-signatures \
+                            --name ${params.STREAM} \
+                            --url https://rhcos.mirror.openshift.com/art/storage/prod/streams \
```
I can add the parameters for `distro` and `url` . But may I know why this part isnt required for others - I didnt quite get that.

mike-nguyen replied:
Both are for future-proofing.  The URL has changed in the past.  For distro we may need to produce images for scos.  They can have default values set to what you currently have here.

aaradhak commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,145 @@
+                        git remote -v
+                        git remote add upstream https://github.com/openshift/installer.git
+                        retries=5
+                        for i in \$(seq 1 \$retries); do
```
I tried with `git fetch upstream release-branch --depth 1` and it seems to fetch without any issue now.

mike-nguyen commented:
Can this job run the `cloud-replicate`  job and create the boot image PR?

aaradhak commented:
> Can this job run the `cloud-replicate` job and create the boot image PR?

This job currently does not support that but it is something that can be worked on from here. At the moment , we would have to run the cloud-replicate job and the bootimagebump jenkins job separately.

aaradhak commented:
The cloud replicate job is now added to the automation , the PR is created after the successful run of cloud-replicate job - https://github.com/openshift/installer/pull/10469

cloud replicate build triggered by bootimagebump job - https://jenkins-rhcos--prod-pipeline.apps.int.prod-stable-spoke1-dc-iad2.itup.redhat.com/job/cloud-replicate/39/console

HuijingHei commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,189 @@
+        durabilityHint('PERFORMANCE_OPTIMIZED')
+    ])
+
+    RHCOS_METADATA_FILE = "data/data/coreos/rhcos.json"
```
For 4.22, we have [coreos-rhel-10.json](https://github.com/openshift/installer/blob/release-4.22/data/data/coreos/coreos-rhel-10.json) and [coreos-rhel-9.json](https://github.com/openshift/installer/blob/release-4.22/data/data/coreos/coreos-rhel-9.json) instead, need to add condition to check that.

aaradhak replied:
Thanks for that, I have added that condition.

HuijingHei commented on jobs/bootimagebump.Jenkinsfile:
```diff
@@ -0,0 +1,213 @@
+                   trim: true),
+            string(name: 'COREOS_ASSEMBLER_IMAGE',
+                   description: 'Override the coreos-assembler image to use',
+                   defaultValue: "quay.io/coreos-assembler/coreos-assembler:rhcos-",
```
For 4.19+, we are using `quay.io/coreos-assembler/coreos-assembler:rhel-9.6`, maybe you can add this into description.

HuijingHei reviewed (approved):
LGTM

aaradhak committed:
bootimagebump: add dual-stream support for 4.22+

For OCP 4.22+, both RHEL 9 and RHEL 10 metadata files need to be
updated. Adding SECONDARY_STREAM and SECONDARY_BUILD_VERSION parameters
to run cloud-replicate and plume cosa2stream for both streams creating
two separate commits in a single PR.

aaradhak commented:
I have added another commit as part of supporting the bootimage bumps in case of dual streams. can I get a review for both the commits. 

cc: @HuijingHei

aaradhak commented:
https://github.com/openshift/installer/pull/10482 created this PR using this automation script

prestist reviewed (approved):
From my perspective it LGTM, though I am not as familiar as I could be with this.

tlbueno reviewed (approved):
lgtm
